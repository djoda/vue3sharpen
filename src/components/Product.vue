<template>
  <div v-if="ratesLoaded" class="card product">
    <div class="card-body">
      <img :src="product.Image" class="card-img-top" alt="..." />
      <div>
        <ul class="list-group list-group-flush ml-3">
          <li class="list-group-item">UPC : {{ product.UPC }}</li>
          <li class="list-group-item">Name {{ product.Name }}</li>
          <li class="list-group-item">
            Price : {{ product.Price.toFixed(2) }} {{ product.Currency }}
          </li>
          <li class="list-group-item">Tax rate : {{ product.TaxRate }}%</li>
          <li class="list-group-item">
            Discount {{ product.calculateDiscount() }} {{ product.Currency }}
          </li>
          <li class="list-group-item">
            Total : {{ product.calculateTotal() }} {{ product.Currency }}
          </li>
        </ul>

        <ul id="divExpenses" class="list-group list-group-flush ml-3">
          <li class="list-group-item">Expenses:</li>
          <li
            class="list-group-item"
            v-for="(expense, index) in this.$store.getters.getExpensesByUPC(
              product.UPC
            )"
            :key="index"
          >
            Description: {{ expense.Description }} Amount: {{ expense.Amount
            }}{{ expense.AmountType }}
          </li>
        </ul>
      </div>

      <div id="btnDiv">
        <button
          id="newTaxBtn"
          class="btn btn-primary"
          data-bs-target="#myModal"
          data-bs-toggle="modal"
          @click="newTaxRate = product.TaxRate"
        >
          New tax rate
        </button>
        <button
          class="btn btn-primary"
          data-bs-target="#expensesModal"
          data-bs-toggle="modal"
        >
          Add expenses
        </button>
        <button
          class="btn btn-primary"
          data-bs-target="#capModal"
          data-bs-toggle="modal"
        >
          Add cap
        </button>
        <label class="d-block select">Currency: </label>
        <select v-model="curr" class="form-control select">
          <option v-for="curr in $store.state.currencies" :key="curr">
            {{ curr }}
          </option>
        </select>
      </div>
    </div>
  </div>

  <div class="modal" id="myModal">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <h5 class="modal-title">New tax rate</h5>
          <button
            type="button"
            class="btn-close"
            data-bs-dismiss="modal"
            aria-label="Close"
          ></button>
        </div>
        <div class="modal-body">
          <label>Set new tax rate for product</label>
          <input
            id="newTaxRateInput"
            type="text"
            class="form-control mt-2"
            v-model.number="newTaxRate"
          />
        </div>
        <div class="modal-footer">
          <button
            type="button"
            class="btn btn-secondary"
            data-bs-dismiss="modal"
          >
            Close
          </button>
          <button
            id="newTaxRateBtnSave"
            type="button"
            class="btn btn-primary"
            @click="updateProduct"
          >
            Save changes
          </button>
        </div>
      </div>
    </div>
  </div>

  <div class="modal" id="expensesModal">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <h5 class="modal-title">Expense</h5>
          <button
            type="button"
            class="btn-close"
            data-bs-dismiss="modal"
            aria-label="Close"
          ></button>
        </div>
        <div class="modal-body">
          <label>Description</label>
          <input type="text" class="form-control mt-2" v-model="description" />
          <label>Set new tax rate for product</label>
          <input
            type="text"
            class="form-control mt-2"
            v-model.number="amount"
          />
          <label>Amount type</label>
          <select class="form-control" v-model="amountType">
            <option value="%">%</option>
            <option value="$">$</option>
          </select>
        </div>
        <div class="modal-footer">
          <button
            type="button"
            class="btn btn-secondary"
            data-bs-dismiss="modal"
          >
            Close
          </button>
          <button type="button" class="btn btn-primary" @click="addExpense">
            Save changes
          </button>
        </div>
      </div>
    </div>
  </div>

  <div class="modal" id="capModal">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-header">
          <h5 class="modal-title">Cap</h5>
          <button
            type="button"
            class="btn-close"
            data-bs-dismiss="modal"
            aria-label="Close"
          ></button>
        </div>
        <div class="modal-body">
          <label>Amount</label>
          <input type="text" class="form-control mt-2" v-model="cap" />
          <label>Cap type</label>
          <select class="form-control" v-model="capType">
            <option value="%">%</option>
            <option value="$">$</option>
          </select>
        </div>
        <div class="modal-footer">
          <button
            type="button"
            class="btn btn-secondary"
            data-bs-dismiss="modal"
          >
            Close
          </button>
          <button type="button" class="btn btn-primary" @click="addCap">
            Save changes
          </button>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { useStore, mapState } from "vuex";
import { useRoute } from "vue-router";
import Expenses from "@/store/Expenses.js";

export default {
  name: "Product",
  setup() {
    let store = useStore();
    let product = store.getters.getByUPC(Number(useRoute().params.upc));

    return {
      product,
    };
  },
  data() {
    return {
      newTaxRate: 20,
      amount: 0,
      description: "",
      amountType: "%",
      cap: 0,
      capType: "%",
      curr: "",
    };
  },
  watch: {
    curr(newValue) {
      if (this.$store.getters.getCurrencyRate(newValue) !== undefined) {
        this.product.Price =
          this.product.PriceUSD * this.$store.getters.getCurrencyRate(newValue);
        this.product.Currency = newValue;
        this.$store.dispatch("updateProduct", this.product);
      }
    },
  },
  methods: {
    updateProduct() {
      this.product.TaxRate = this.newTaxRate;
      this.$store.dispatch("updateProduct", this.product);
    },
    addExpense() {
      if (this.amount > 0) {
        this.$store.dispatch(
          "addExpense",
          new Expenses(
            this.description,
            this.amount,
            this.amountType,
            this.product.UPC
          )
        );
      }
      this.amount = 0;
      this.description = "";
    },
    addCap() {
      this.product.Cap = this.cap;
      this.product.CapType = this.capType;
      this.$store.dispatch("updateProduct", this.product);
    },
  },
  mounted() {
    this.curr = this.product.Currency;
  },
  computed: {
    ...mapState(["ratesLoaded"]),
  },
};
</script>

<style scoped>
.product {
  display: flex;
  flex: 0.3 1 500px;
  height: 50vh;
  margin-top: 50px;
}

.product img {
  width: 200px;
  height: 200px;
  position: absolute;
  top: 5px;
  left: 35px;
}

.product ul {
  position: absolute;
  top: 215px;
  left: 35px;
}

#btnDiv {
  position: absolute;
  top: 15px;
  left: 260px;
}

#btnDiv button {
  margin-left: 10px;
}

#divExpenses {
  position: absolute;
  top: 215px;
  left: 260px;
}

.select {
  margin: 10px;
}
</style>