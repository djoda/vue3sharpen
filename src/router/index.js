import { createRouter, createWebHistory } from 'vue-router'
import Products from "@/components/Products.vue"
import Discount from "@/components/Discount.vue"

const routes = [
  {
    path: "/",
    name: "Products",
    component: Products
  },
  {
    path: "/discount",
    name: "Discount",
    component: Discount
  }
]

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes
})

export default router
