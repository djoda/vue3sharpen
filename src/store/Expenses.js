class Expenses {
    constructor(description, amount, amonutType, productUPC) {
        this.Description = description;
        this.Amount = amount;
        this.AmountType = amonutType;
        this.ProductUPC = productUPC;
    }
} export default Expenses;