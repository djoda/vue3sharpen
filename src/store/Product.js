import { useStore } from "vuex";

class Product {
    constructor(name, UPC, price, image) {
        this.Name = name;
        this.UPC = UPC;
        this.Price = price;
        this.TaxRate = 20;
        this.UPCDiscount = 0;
        this.Image = image;
    }

    getUniversalDiscount() {
        return (useStore()).state.universalDiscount;
    }

    calculateGlobal() {
        return this.Price * this.getUniversalDiscount() / 100;
    }

    calculateDiscount() {
        if (useStore().state.additiveMode && this.UPCDiscount != 0) {
            return this.calculateGlobal() + this.calculateUCPDiscount();
        } else {
            return (this.Price - this.calculateGlobal()) * this.UPCDiscount / 100;
        }
    }

    calculateTax() {
        // console.log((useStore()).state.beforeDiscount)
        if ((useStore()).state.beforeDiscount)
            return this.Price * (this.TaxRate / 100);
        else {
            // console.log("ASDAS", this.calculateDiscount(this.getUniversalDiscount()))
            return (this.Price - this.calculateDiscount()) * this.TaxRate / 100;
        }
    }

    calculateUCPDiscount() {
        return this.Price * this.UPCDiscount / 100;
    }

    calculateExpenses() {
        let total = 0;
        for (let expense of useStore().state.expenses) {
            if (expense.AmountType === "%") {
                total += this.Price * expense.Amount / 100;
            } else {
                total += expense.Amount;
            }
        }
        return total;
    }

    calculateTotal() {
        return (
            (this.Price + this.calculateExpenses() - this.calculateDiscount() + this.calculateTax()).toFixed(2)
        );
    }
}
export default Product;