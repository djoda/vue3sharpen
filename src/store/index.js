import { createStore } from 'vuex'
import Product from "./Product"
import saveToLocalStorage from "./Plugins/webStorage"
import Expenses from './Expenses';

export default createStore({
  state: {
    mockProducts: [],
    products: [new Product("Product1", 111, 4565, "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"),
    new Product("Product2", 222, 3622, "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"),
    new Product("Product3", 333, 441565, "https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg"),
    new Product("Product4", 444, 56, "https://fakestoreapi.com/img/71pWzhdJNwL._AC_UL640_QL65_ML3_.jpg")],
    universalDiscount: 0,
    expenses: [new Expenses("Transport", 500, "$", 111)]
  },
  getters: {
    getByUPC: (state) => (UPC) => {
      return state.products.find(e => e.UPC === UPC);
    },
    getExpensesByUPC: (state) => (UPC) => {
      return state.expenses.filter(e => e.ProductUPC == UPC);
    }
  },
  mutations: {
    addProduct(state, item) {
      state.products.push(item);
    },
    setUniversalDiscount(state, ud) {
      state.universalDiscount = ud;
    },
    setMockProducts(state, items) {
      state.mockProducts = items;
    },
    setUPCDiscount(state, payload) {
      state.products = [...state.products.filter(e => e.UPC != payload.UPC), payload];
    },
    addExpense(state, payload) {
      state.expenses.push(payload);
      console.log(state.expenses)
    }
  },
  actions: {
    async fetchPosts({ commit }) {
      const products = await fetch("https://fakestoreapi.com/products").then(res => res.json());
      commit("setMockProducts", products);
    },
    async updateUPCDiscount(context, payload) {
      let prod = context.getters.getByUPC(payload.UPC);
      prod.UPCDiscount = payload.UPCDiscount;
      context.commit("setUPCDiscount", prod);
    },
    async updateProduct(context, payload) {
      context.commit("addProduct", payload);
    },
    async addExpense(context, payload) {
      context.commit("addExpense", payload);
    }
  },
  modules: {
  },
  plugins: [saveToLocalStorage]
})

