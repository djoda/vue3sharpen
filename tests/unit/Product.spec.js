import { mount, } from "@vue/test-utils"
import Product from "@/components/Product.vue"
import store from "@/store"
import router from "@/router"
import { config } from '@vue/test-utils'
import ProductTest from "../ProductTest"


describe("Product.vue", () => {
    test("snapshot", () => {
        const wrapper = mount(Product, {
            global: {
                plugins: [store, router],
                mocks: {
                    product: new ProductTest("Product1", 111, 4565, "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"),
                }
            },
        });
        expect(wrapper.element).toMatchSnapshot();
    })
})