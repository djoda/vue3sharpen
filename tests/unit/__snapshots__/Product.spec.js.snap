// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product.vue snapshot 1`] = `
<div
  data-v-app=""
>
  
  <!--v-if-->
  <div
    class="modal"
    id="myModal"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <h5
            class="modal-title"
          >
            New tax rate
          </h5>
          <button
            aria-label="Close"
            class="btn-close"
            data-bs-dismiss="modal"
            type="button"
          />
        </div>
        <div
          class="modal-body"
        >
          <label>
            Set new tax rate for product
          </label>
          <input
            class="form-control mt-2"
            type="text"
          />
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn btn-secondary"
            data-bs-dismiss="modal"
            type="button"
          >
             Close 
          </button>
          <button
            class="btn btn-primary"
            type="button"
          >
             Save changes 
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="modal"
    id="expensesModal"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <h5
            class="modal-title"
          >
            Expense
          </h5>
          <button
            aria-label="Close"
            class="btn-close"
            data-bs-dismiss="modal"
            type="button"
          />
        </div>
        <div
          class="modal-body"
        >
          <label>
            Description
          </label>
          <input
            class="form-control mt-2"
            type="text"
          />
          <label>
            Set new tax rate for product
          </label>
          <input
            class="form-control mt-2"
            type="text"
          />
          <label>
            Amount type
          </label>
          <select
            class="form-control"
          >
            <option>
              %
            </option>
            <option>
              $
            </option>
          </select>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn btn-secondary"
            data-bs-dismiss="modal"
            type="button"
          >
             Close 
          </button>
          <button
            class="btn btn-primary"
            type="button"
          >
             Save changes 
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="modal"
    id="capModal"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <h5
            class="modal-title"
          >
            Cap
          </h5>
          <button
            aria-label="Close"
            class="btn-close"
            data-bs-dismiss="modal"
            type="button"
          />
        </div>
        <div
          class="modal-body"
        >
          <label>
            Amount
          </label>
          <input
            class="form-control mt-2"
            type="text"
          />
          <label>
            Cap type
          </label>
          <select
            class="form-control"
          >
            <option>
              %
            </option>
            <option>
              $
            </option>
          </select>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn btn-secondary"
            data-bs-dismiss="modal"
            type="button"
          >
             Close 
          </button>
          <button
            class="btn btn-primary"
            type="button"
          >
             Save changes 
          </button>
        </div>
      </div>
    </div>
  </div>
  
</div>
`;
